drivers/platform/aducm3029/trng.c

```diff
@@ -0,0 +1,169 @@
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <stdlib.h>
+#include <drivers/rng/adi_rng.h>
+#include "trng.h"
+#include "util.h"
+#include "error.h"
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+
+/* LenReload: 0 - 4095 */
+#define TRNG_CNT_VAL    4095
+/* Prescaler: 0 - 10 */
+#define TRNG_PRESCALER  2
+
+/******************************************************************************/
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+/**
+ * @struct trng_desc
+ * @brief Stucture holding the TRNG descriptor.
+ */
+struct trng_desc {
+	/** Memory used by the DFP */
+	uint32_t	dev_mem[(ADI_RNG_MEMORY_SIZE + 3)/4]; //Aligned memory
+	/** DFP Hanler */
+	ADI_RNG_HANDLE	dev;
+};
+
+/******************************************************************************/
+/**************************** Global Variables ********************************/
+/******************************************************************************/
+
+/** Number of references to the descriptor */
+static uint32_t		nb_references;
+/** TRNG Descriptor */
+static struct trng_desc	g_desc;
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/* Initialize the TRNG device and the global descriptor */
+static void gdesc_init()
+{
+	adi_rng_Open(0, g_desc.dev_mem, ADI_RNG_MEMORY_SIZE, &g_desc.dev);
+
+	adi_rng_SetSampleLen(g_desc.dev, TRNG_PRESCALER, TRNG_CNT_VAL);
+
+	adi_rng_EnableBuffering(g_desc.dev, false);
+
+	adi_rng_Enable(g_desc.dev, true);
+}
+
+/* Remove the TRNG device and the global descriptor */
+static void gdesc_remove()
+{
+	adi_rng_Enable(g_desc.dev, false);
+	adi_rng_Close(g_desc.dev);
+}
+
+/**
+ * @brief Initialize descriptor
+ * @param desc - Where to store the TRNG Descriptor
+ * @return \ref SUCCESS or \ref FAILURE if desc is null
+ */
+int32_t trng_init(struct trng_desc **desc)
+{
+
+	if (!desc)
+		return FAILURE;
+
+	if (!nb_references)
```

Comment:
Could nb_references be moves into some sort of a descriptor instead of being global ?

Reply:
No, nb_reference is used to count the references to the global descriptor